Feature/user

leejaeeun59357 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
프로젝트 시작

**TO-BE**
- Security & Jwt
  - 세션 방식을 사용하지 않고 토큰 방식 사용

- 회원가입 기능
  - 회원가입 시 아이디와 패스워드, 이메일을 입력받는다.
  - 이때, 아이디와 이메일은 중복될 수 없다.
  - 입력받은 아이디를 이용하여 token 생성하고 이메일로 토큰 정보를 포함한 인증링크를 전송
  - 인증이 완료되면 verified_status true로 변환
  - DB에 사용자정보 저장 시 패스워드는 PasswordEncoder를 사용하여 암호화한다.

- 로그인 기능
  - ID와 패스워드를 통해 사용자가 있는지 확인 후 JWT Token 발행

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] 이메일 정상적으로 전송되는지 테스트
- [x] 이메일으로 전송된 링크 클릭 시 사용자 인증 상태 변화하는지 테스트
- [x] 로그인 정상적으로 되는지 테스트

ne2227 committed:
[chore] build.gradle 변경

ne2227 committed:
[feat] 회원가입 기능 구현
- 회원가입 시 아이디, 패스워드, 이메일을 입력받음
- 아이디와 이메일이 중복되면 에러 발생
- 회원가입 시 회원 권한 'USER'으로 설정
- 아직 이메일 인증을 받지 못했으므로 verified_status는 false 로 DB에 저장
- 회원가입 테스트 코드 작성 (scratch_user.http)

ne2227 committed:
[feat] 회원가입 시 입력받은 패스워드 암호화 기능 추가
- spring security를 사용하여 비밀번호를 암호화하여 DB에 저장하는 기능을 추가함
- 입력받은 email이 DB저장까지 넘어가지 않는 부분 수정

ne2227 committed:
[feat] 회원가입 시 이메일 인증을 위한 Token 생성 기능 구현
- SignupForm을 통해 입력받은 이메일의 인증을 위해 토큰의 Entity를 생성 및 저장
- SignupForm을 통해 받은 loginId를 이용하여 토큰 생성

ne2227 committed:
[feat] Token의 @Id 를 생성 및 uuid2 형식의 String으로 생성
- 인증 링크 전송 시 token을 식별할 수 있는 id가 필요하기 때문에 외래키를 기본키로 사용하지 않고 기본키를 따로 생성함

ne2227 committed:
[feat] 회원가입 시 입력받은 email로 인증 링크 전송하는 기능 구현
- 회원가입 시 동시에 인증 링크를 발송하는 기능 구현
- 인증링크에 해당 사용자의 id를 사용하여 생성된 token의 값을 포함시킴

ne2227 committed:
[refactor] 사용하지 않는 변수 삭제 및 주석 작성

ne2227 committed:
[feat] 사용자에게 전송된 인증 링크를 통해 이메일 인증 기능 구현

- userId를 통해 생성된 tokenId를 인증링크에 포함시켜 전송
- 인증이 성공하면 user의 verified_status를 true로 변경 및 저장
- 인증 시 만료시간이 지났다면 Exception 발생
- 인증 성공 시, String 출력

ne2227 committed:
[refactor] 이메일 인증 시 예외 추가

- 이미 인증이 완료된 이메일이라면 Exception 발생

ne2227 committed:
[refactor] SecurityConfig 파일 삭제 및 재생성

ne2227 committed:
Merge remote-tracking branch 'origin/feature/User' into feature/User

# Conflicts:
#	src/main/java/org/board/exercise_board/User/Security/SecurityConfig.java

ne2227 committed:
[refactor] SecurityConfig 파일 삭제 및 재생성

ne2227 committed:
[feat] 로그인 기능 구현

- JWT를 사용한 로그인 기능 구현
- 로그인 테스트용 scratch 파일 작성

je-pa commented on src/main/java/org/board/exercise_board/User/domain/model/Token.java:
```diff
@@ -0,0 +1,40 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@AuditOverride(forClass = BaseEntity.class)
+public class Token extends BaseEntity{
```
토큰을 DB에 저장하는 걸까요? 
혹시 왜 저장하는지 알 수 있을까요?

leejaeeun59357 replied:
해당 사용자의 정보로 token을 생성하고 인증링크에 해당 token 정보를 담아서 보내도록 하고 있습니다!
이 때 해당 token이 어떤 사용자의 것인지 파악하기 위해 token을 db에 저장하고 있습니다!

je-pa commented on src/main/java/org/board/exercise_board/User/service/UserService.java:
```diff
@@ -0,0 +1,69 @@
+   * @return true/false
+   */
+  public boolean isExistEmail(String email) {
+    return userRepository.findByEmail(email).isPresent();
```
jpa로 repository에 existsByEmail 선언해서 써도 괜찮을 것 같아요!

leejaeeun59357 replied:
옙! 수정하겠습니다!!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/Security/CustomUserDetails.java:
```diff
@@ -0,0 +1,52 @@
+@RequiredArgsConstructor
+public class CustomUserDetails implements UserDetails {
+
+  private final User user;
```
CustomUserDetail에서 User를 그대로 field로 들고 있다면 entity Manager가 CustomUserDetails가 존재하는 곳까지 있어야할 것 같아요. CustomUserDetail에서 User type을 들고 있기 보다는 필요한 필드만 들고 있도록 변경되면 좋을것 같습니다

leejaeeun59357 replied:
제 생각에는 해당 User의 정보를 사용해야 하기 때문에 User를 사용하도록 했는데 혹시 별도의 dto 를 생성해서 사용하는것이 멘토님이 말씀해주시는 내용인지 궁금합니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/domain/Dto/UserDto.java:
```diff
@@ -0,0 +1,48 @@
+
+  private String loginId;
+  private String email;
+  private Boolean verified_status;
```
java convention에 맞춰서 카멜케이스로 작성해주시면 좋을 것 같습니다

leejaeeun59357 replied:
넵 수정하겠습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/domain/Dto/UserDto.java:
```diff
@@ -0,0 +1,48 @@
+   * @param signUpForm
+   * @return
+   */
+  public static User formToEntity(SignUpForm signUpForm) {
+    return User.builder()
+        .loginId(signUpForm.getLoginId())
+        .email(signUpForm.getEmail())
+        .password(signUpForm.getPassword())
+        .verified_status(false)
+        .role(Role.USER)
+        .build();
+  }
```
signUpForm을 User로 변경되는 코드인데 UserDto에 존재하고 있네요. SignUpForm 클래스나 나 User 클래스 아래에 존재하면 좋을 것 같아요

leejaeeun59357 replied:
넵! 수정하겠습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/domain/model/User.java:
```diff
@@ -0,0 +1,44 @@
+  @Column(unique = true)
+  private String email;
+
+  private Role role;
```
@Enumerated를 안붙여줘도 괜찮을까요

leejaeeun59357 replied:
수정하겠습니다!!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/domain/model/User.java:
```diff
@@ -0,0 +1,44 @@
+
+  // 이메일 인증이 완료되면 true로 변경하기 위해 @Setter 사용
+  @Setter
+  private Boolean verified_status;
```
마찬가지로 카멜케이스로 작성되면 좋을 것 같아요.

leejaeeun59357 replied:
수정하겠습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/domain/model/Token.java:
```diff
@@ -0,0 +1,40 @@
+  private User user;
+
+  @Setter
+  private LocalDateTime expirationDate;
```
필드가 dateTime이라면 네이밍도 dateTime이면 좋을 것 같아요. 시간관계없이 일시만 저장한다면 LocalDate를 이용해도 좋구요

leejaeeun59357 replied:
수정하겠습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/service/EmailService.java:
```diff
@@ -0,0 +1,41 @@
+      javaMailSender.send(message);
+
+    } catch (Exception e) {
+      log.info(e.toString());
```
이메일을 보내다가 에러가 났을 경우 다시 에러를 던져주지 않고 info로 로그를 찍어도 괜찮은가요?
이메일을 보내다가 에러가 난경우 서버에서 처리해주어야할 작업은 없을까요? (재시도라던가, 유저에게 알려준다던가)

leejaeeun59357 replied:
log 출력이 아니라 에러 발생하는 것으로 수정하였습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/service/TokenService.java:
```diff
@@ -0,0 +1,55 @@
+  private final TokenRepository tokenRepository;
+
+  // 만료 시간은 5분으로 설정
+  private final long EMAIL_TOKEN_EXPIRATION_TIME_VALUE = 5L;
```
시간의 단위도 같이 써주시면 좋을 것 같아요. 현재는 분이 단위니까
`EMAIL_TOKEN_EXPIRATION_MINUTE_TIME_VALUE` 이런식으로요

leejaeeun59357 replied:
넵! 수정하겠습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/service/TokenService.java:
```diff
@@ -0,0 +1,55 @@
+   */
+  public boolean verifyExpirationDate(Token token) {
+    // 현재 시간이 만료시간을 지나지 않았을 때 true
+    if(LocalDateTime.now().isBefore(token.getExpirationDate()) ||
+    LocalDateTime.now().isEqual(token.getExpirationDate())) {
+      return true;
+    } else {
+      return false;
+    }
```
`return !LocalDateTime.now().isAfter(token.getExpirationDate())`
이렇게 간편화시킬수도 있을 것 같아요.

leejaeeun59357 replied:
넵 수정하겠습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/service/TokenService.java:
```diff
@@ -0,0 +1,55 @@
+   */
+  public void updateVerifyStatus(Token token) {
+    tokenRepository.save(token);
+  }
```
메소드명은updateVerifyStatus인데 token의 verifyStatus를 변경하는 작업은 이 메소드에서 진행하지 않는 것 같아요.
메소드 자체는 save 로직만 수행하는 것으로 보이네요.
메소드명과 동작하는 방식이 일치하면 좋을 것 같아요.

leejaeeun59357 replied:
예 수정하겠습니다!!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/service/TokenService.java:
```diff
@@ -0,0 +1,55 @@
+    Token token = new Token();
+    token.setUser(user);
+    token.setExpirationDate(LocalDateTime.now().plusMinutes(EMAIL_TOKEN_EXPIRATION_TIME_VALUE));
+    return tokenRepository.save(token);
```
user가 토큰을 생성후, 토큰 만료, 다시 createToken을 호출했을때 이 코드에서 따로 문제는 없을까요? 새롭게 생성해서 token을 save하고 있어 기존에 저장된 데이터와 merge가 잘 되는지 궁금합니다

leejaeeun59357 replied:
만약 존재한다면 해당 토큰을 사용하도록 수정하였습니다!

mingming-mentor reviewed (approved):
리뷰드렸습니다. 수고 많으셨어요

ne2227 committed:
[refactor] 변수명 수정 및 코드 간소화

- 변수명을 카멜형식으로 변경
- 코드 위치 변경 (UserDto -> User)
- 코드 간소화
- 사용하지 않는 import문 삭제

ne2227 committed:
[feat] 코드 일부 수정

- 메일 전송 실패 시 에러 발생으로 변경
- 토큰 생성 시 해당 사용자에 대한 token 정보가 존재하는지 확인하고 존재한다면 덮어쓰기로 변경
